### PR TITLE
ConstExpr improvements for strings and moving towards arrays.

### PR DESCRIPTION
**Pull request by lattner**
Improve ConstExpr processing for strings to handle ones with literals correctly.
Start treating Builtin.RawPointer as addresses, which will be necessary for array
literals.

Refactor the code that analyzes calls to _allocateUninitializedArray and move it into
ConstExprEvaluator (not the best place perhaps, but better than what we had before).

Add a testcase showing that we handle empty and non-empty strings now.

### CODE REVIEW

**Comment by lattner** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/TFConstExpr.cpp [hunk 1]
```diff
@@ -1557,3 +1642,133 @@ computeConstantValues(ArrayRef<SILValue> values,
...
+    // Oops we found an unknown use!
+    return true;
+  }
+  return false;
+}
+
+/// Try to decode the specified apply of the _allocateUninitializedArray
+/// function in the standard library.  This attempts to figure out what the
+/// resulting elements will be.  This fills in the elements result and returns
+/// true on success.
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFConstExpr.cpp]: some other existing functions return "true" on failure. 

My first preference is to "return true on success" consistently throughout the code base, but i understand that can involve changing large parts of the code base and may not be worthwhile.

My fallback preference is to "return true on failure" for consistency reasons.

What do you think?

File: lib/SILOptimizer/Mandatory/TFConstExpr.h [hunk 2]
```diff
@@ -61,6 +61,20 @@ class ConstExprEvaluator {
...
+  /// true on success.
+  ///
+  /// If arrayInsts is non-null and if decoding succeeds, this function adds
+  /// all of the instructions relevant to the definition of this array into
+  /// the set.  If decoding fails, then the contents of this set is undefined.
+  ///
+  static bool decodeAllocUninitializedArray(ApplyInst *apply,
+                                            uint64_t numElements,
+                                            SmallVectorImpl<SILValue> &elements,
+                                   SmallPtrSet<SILInstruction*, 8> *arrayInsts);
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFConstExpr.h]: align this param

File: lib/SILOptimizer/Mandatory/TFConstExpr.cpp [hunk 3]
```diff
@@ -838,16 +858,74 @@ ConstExprFunctionState::computeCallResult(ApplyInst *apply) {
...
+    for (auto elt : elements) {
+      auto eltCst = getConstantValue(elt);
+      if (!eltCst.isConstant())
+        return eltCst;
+      elementConstants.push_back(eltCst);
+    }
+
+    auto v = SymbolicValue::getAggregate(elementConstants,
+                                         evaluator.getAllocator());
+    //llvm::errs() << "Got it: " << v;
```
Comment (mhong) [inline lib/SILOptimizer/Mandatory/TFConstExpr.cpp]: remove this or wrap with DEBUG()?

**Review by mhong — APPROVED** [review]
Nice!

File: lib/SILOptimizer/Mandatory/TFConstExpr.cpp (hunk 1 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/TFConstExpr.cpp]: +1

File: lib/SILOptimizer/Mandatory/TFConstExpr.h (hunk 2 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/TFConstExpr.h]: I believe this is for fitting within 80 columns.

File: lib/SILOptimizer/Mandatory/TFConstExpr.cpp (hunk 1 above)
Comment (lattner) [reply lib/SILOptimizer/Mandatory/TFConstExpr.cpp]: Sure, will change.

File: lib/SILOptimizer/Mandatory/TFConstExpr.h (hunk 2 above)
Comment (lattner) [reply lib/SILOptimizer/Mandatory/TFConstExpr.h]: Right, I flowed it a different way, it is a bit less weird.

File: lib/SILOptimizer/Mandatory/TFConstExpr.cpp (hunk 3 above)
Comment (lattner) [reply lib/SILOptimizer/Mandatory/TFConstExpr.cpp]: This is incomplete, I need to build this out when I have a way to represent arrays.  Still WIP.  I'll #if 0 is out :-)

**Comment by lattner** [comment]
Thanks for the feedback, I need to fix an issue handling StaticString that I regressed.  I'll try to do that later tonight or tomorrow.

**Comment by lattner** [comment]
@swift-ci please test tensorflow

**Comment by lattner** [comment]
@swift-ci please test tensorflow

**Comment by lattner** [comment]
@swift-ci please test tensorflow

**Comment by lattner** [comment]
@swift-ci please test tensorflow

**Comment by lattner** [comment]
@swift-ci please test tensorflow

**Comment by lattner** [comment]
Oh, sorry @rxwei I missed your comment.  Please still do take a look, I'll incorporate any improvements into my next patch.  I need to change a lot of the internals... sigh.

**Pull request merged**